pkg/vgmanager/lvm.go

```diff
@@ -23,33 +23,218 @@ import (
 	"github.com/red-hat-storage/lvm-operator/pkg/internal"
 )
 
+type lvmError string
+
+func (s lvmError) Error() string { return string(s) }
+
+var (
+	ErrVolumeGroupNotFound  = lvmError("volume group not found")
+	ErrNoNewPhysicalVolumes = lvmError("no new physical volumes")
+)
+
 const (
-	lvmPath = "/usr/sbin/lvm"
+	lvmPath     = "/usr/sbin/lvm"
+	extendVGCmd = "/usr/sbin/vgextend"
+	createVGCmd = "/usr/sbin/vgcreate"
+	vgRemoveCmd = "/usr/sbin/vgremove"
+	pvRemoveCmd = "/usr/sbin/pvremove"
 )
 
+// vgsOutput represents the output of the `vgs --reportformat json` command
+type vgsOutput struct {
+	Report []struct {
+		Vg []struct {
+			Name string `json:"vg_name"`
+		} `json:"vg"`
+	} `json:"report"`
+}
+
+// pvsOutput represents the output of the `pvs --reportformat json` command
+type pvsOutput struct {
+	Report []struct {
+		Pv []struct {
+			Name   string `json:"pv_name"`
+			VgName string `json:"vg_name"`
+		} `json:"pv"`
+	} `json:"report"`
+}
+
 // VolumeGroup represents a volume group of linux lvm.
 type VolumeGroup struct {
+	// Name is the name of the volume group
 	Name string `json:"vg_name"`
+
+	// PVs is the list of physical volumes associated with the volume group
+	PVs []string `json:"pvs"`
 }
 
-// ListVolumeGroups lists all volume groups.
-func ListVolumeGroups(exec internal.Executor) ([]VolumeGroup, error) {
-	// volumeGroupMap type is used for marshalling the `vgs --reportformt json` command output
-	// Sample output:
-	// `{"report": [ {"vg": [{"vg_name":"fedora", "pv_count":"1", "lv_count":"3", "snap_count":"0", "vg_attr":"wz--n-", "vg_size":"<475.94g", "vg_free":"0 "}]}]`
-	volumeGroupMap := map[string][]map[string][]VolumeGroup{}
-	arg := []string{
+// Create creates a new volume group
+func (vg VolumeGroup) Create(exec internal.Executor, pvs []string) error {
+	if vg.Name == "" {
+		return fmt.Errorf("failed to create volume group. Volume group name is empty")
+	}
+
+	if len(pvs) == 0 {
+		return fmt.Errorf("failed to create volume group. Physical volume list is empty")
+	}
+
+	args := []string{vg.Name}
+	args = append(args, pvs...)
+
+	_, err := exec.ExecuteCommandWithOutputAsHost(createVGCmd, args...)
+	if err != nil {
+		return fmt.Errorf("failed to create volume group %q. %v", vg.Name, err)
+	}
+
+	return nil
+}
+
+// Extend extends the volume group only if new physical volumes are available
+func (vg VolumeGroup) Extend(exec internal.Executor, pvs []string) error {
+	// Extend VG only if there is any new Physical Volume
+	newPVs := listDiff(pvs, vg.PVs)
```

Comment:
`filterMatchingDevices` would ensure that `pvs` list does not contains PVs that does not belong to other VGs. `listDiff` will ensure that we only run `vgextend` with new set of `PVs`.

Reply:
FilterDevices does not list PVs. It is returning a list of devices only. However, this function is comparing PVs to the PVs in the VG. Based on the current code, the presence of a PV would mean that it is already being used in a VG (as it does not create them explicitly but as part of a vgcreate/vgextend call) so comparing it to a particular VGs PV list is incorrect and unnecessary.

Reply:
vgextend command should always contain new physical volumes, else it will fail.  This is a bug in the current code as well. For example: 
``` 
# lsblk
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda      8:0    0  19.5G  0 disk 
`-sda1   8:1    0  19.5G  0 part /mnt/sda1
sdb      8:16   0    10G  0 disk 
sdc      8:32   0     8G  0 disk 
sr0     11:0    1 225.6M  0 rom  
# vgs
# vgcreate vg1 /dev/sdb
  Physical volume "/dev/sdb" successfully created.
  Volume group "vg1" successfully created
# vgs
  VG  #PV #LV #SN Attr   VSize   VFree  
  vg1   1   0   0 wz--n- <10.00g <10.00g
# vgextend vg1 /dev/sdb  /dev/sdc
  Physical volume '/dev/sdb' is already in volume group 'vg1'
  Unable to add physical volume '/dev/sdb' to volume group 'vg1'
  /dev/sdb: physical volume not initialized.
# vgs
  VG  #PV #LV #SN Attr   VSize   VFree  
  vg1   1   0   0 wz--n- <10.00g <10.00g
# 
``` 
`listdiff` tries to ensure that `vgextend` will always use new Physical volumes.

Reply:
removed `istdiff`